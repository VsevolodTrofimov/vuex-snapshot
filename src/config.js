/**
 * @namespace 
 * @property {Boolean} autoResolve resolve all MockPromises and fetches in order they were created
 * @property {Boolean} snapEnv include state, getters and paylaod into snapshot
 */
const options = {
  autoResolve: false,
  snapEnv: false
}

// they are likely to stay flat
const defaults = Object.assign({}, options)
const reset = () => Object.assign(options, defaults)
reset()

export default {
  options,
  reset
}