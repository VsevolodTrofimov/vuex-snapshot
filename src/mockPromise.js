import timetable from './timetable'
import {useGlobally} from './utility'


const RealPromise = Promise


export class MockPromise extends RealPromise {
  /**
   * Named promise that can be resolved manually and properly serialized 
   * and registers it in timetable
   * @param {Function} cb 
   * @param {string} name
   */
  constructor(cb, name='Promise') {
    let resovleTrigger
    let rejectTrigger

    // name-only construction
    if(typeof cb === 'string') {
      name = cb
      cb = () => {}
    }

    const cbProxy = (resolve, reject) => {
      resovleTrigger = resolve
      rejectTrigger = reject
      cb(resolve, reject)
    }

    super(cbProxy)

    this.name = name
    this.resolve = resovleTrigger
    this.reject = rejectTrigger

    timetable.register({
      name,
      promise: this,
      payload: cb,
      resolve: resovleTrigger,
      reject: rejectTrigger
    })
  }
}


export const useMock = () => useGlobally('Promise', MockPromise)
export const useReal = () => useGlobally('Promise', RealPromise)