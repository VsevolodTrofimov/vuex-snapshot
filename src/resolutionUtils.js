import serialize from './serialize'


export const normalizeResolution = resolution => {
  const normalResolution = {
    name: '',
    type: 'resolve',
    payload: undefined,
  }
  
  // string constructor
  if(typeof resolution === 'string' || resolution.indexOf('string')) {
    normalResolution.name = resolution
    return normalResolution
  } 
  
  // object constructor, errors are duplicated because they are likely to occur in promises
  if(resolution.name) {
    normalResolution.name = resolution.name
  } else {
    throw new Error('vuex-snapshot: INPUT ERROR resolution must have a name')
  }

  if(resolution.type && ['resolve', 'reject'].indexOf(resolution.type) !== -1) {
    normalResolution.type = resolution.type
  } else {
    throw new Error('vuex-snapshot: INPUT ERROR resolution type must be' 
                    + 'either "resovle" or "reject"')
  }

  if(resolution.payload) {
    normalResolution.payload = resolution.payload
  }
}


export const simualteResolution = (resolution, snapCb, timetable) => {
  snapCb(`RESOLUTION: ${resolution.name} -> ${resolution.type}`, resolution.payload)
  return timetable.trigger(resolution)
}


export const simualteResolutions = (resolutions, snapCb, timetable) => {
  return new Promise((resolveSimulation, rejectSimulation) => {
    const normalResolutions = resolutions.map(normalizeResolution)
    
    // simulates given resolution and queues the next until all are simulated
    const simulationLoop = idx => {
      if(idx === normalResolutions.length) {
        resolveSimulation()
      }
      else {
        simualteResolution(normalResolutions[idx], snapCb, timetable)
          .then(() => simulationLoop(idx + 1))
          .catch(rejectSimulation)
      }
    }

    simulationLoop(0)
  })
}
