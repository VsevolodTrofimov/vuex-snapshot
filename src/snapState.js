import serialize from './serialize'


/**
 * Serializes your state the fancy way 
 * (including functions and mock promises)
 * @param {*} state 
 */
export const snapState = state => serialize(state)
export default snapState