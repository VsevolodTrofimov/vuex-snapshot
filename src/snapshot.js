class Snapshot {
  constructor() {
    this.value = []
    this.add = this.add.bind(this)
  }
  add(message, payload) {
    const entry = {}
    entry.message = message
    if(typeof payload !== 'undefined') entry.payload = payload
    this.value.push(entry)
  }
}


export default Snapshot