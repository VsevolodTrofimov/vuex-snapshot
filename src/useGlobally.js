const useGlobally = (name, value) => {
  if(typeof window === 'undefined') global[name] = value // node
  else window[name] = value // browser
}

export default useGlobally