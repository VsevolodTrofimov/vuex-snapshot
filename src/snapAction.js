import Snapshot from './snapshot'
import {simualteResolutions} from './resolutionUtils'
import timetable from '../src/timetable'

const RealPromise = Promise


export const makeCallSnapper = (snapshot, type, cb) => (name, payload) => {
  snapshot.add(`${type}: ${name}`, payload)
  cb(name, payload)
}


/**
 * @typedef {{name:string, type: ("resolve" | "reject"), payload}} Resolution
 */
/**
 * Takes snapshot of action's evaluation
 * @param {Function} action action to test
 * @param {{state, getters, commit: Function, dispatch: Function, payload}} mocks arguments passed to the action, payload is the second argument
 * @param {[(string | Resolution)]} resolutions
 * @param {{autoResovle: Boolean, snapEnv: Boolean}} options
 * @returns  {(string | Promise<string>)}
 */
export const snapAction = (action, mocks, resolutions, options) => {
  const snapshot = new Snapshot()
  const mockCommit = makeCallSnapper(snapshot, 'COMMIT', mocks.commit)
  const mockDispatch = makeCallSnapper(snapshot, 'DISPATCH', mocks.dispatch)

  if(options.snapEnv) {
    snapshot.add('DATA MOCKS', {
      state: mocks.state,
      getters: mocks.getters
    })
    snapshot.add('ACTION CALL', mocks.payload)
  }

  const actionReturn = action({
    commit: mockCommit,
    dispatch: mockDispatch,
    state: mocks.state,
    getters: mocks.getters
  }, mocks.payload)

  if(typeof actionReturn !== 'undefined' && actionReturn instanceof Promise) {
    // action is async
    return new RealPromise((resolve, reject) => {
      actionReturn
        .then(payload => {
          snapshot.add('ACTION RESOLVED', payload)
          resolve(snapshot.value)
        })
        .catch(payload => {
          snapshot.add('ACTION REJECTED', payload)
          resolve(snapshot.value)
        })
      
      simualteResolutions(resolutions, snapshot, timetable, options)
        .then(() => {
          // this is needed to let action to resolve first
          setTimeout(() => {
            snapshot.add('ACTION DID NOT RESOLVE')
            resolve(snapshot.value)
          }, 0)
        })
        .catch(err => {
          reject({
            err,
            run: snapshot.value
          })
        })

    })
  } else {
    // action is sync
    return snapshot.value
  }
}

export default snapAction