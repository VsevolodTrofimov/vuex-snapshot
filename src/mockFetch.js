import timetable from './timetable'
import useGlobally from './useGlobally'


const RealPromise = Promise
const realFetch = fetch


// So we can resolve them manually and serialize them
export const mockFetch = (url, init) => {
  let resovleTrigger
  let rejectTrigger

  const cbProxy = (resolve, reject) => {
    resovleTrigger = resolve
    rejectTrigger = reject
  }

  const simulation = new RealPromise(cbProxy)

  timetable.register({
    name: url,
    payload: init,
    resolve: resovleTrigger,
    reject: rejectTrigger
  })

  return simulation
}


export const useMock = () => useGlobally('fetch', mockFetch)
export const useReal = () => useGlobally('fetch', realFetch)