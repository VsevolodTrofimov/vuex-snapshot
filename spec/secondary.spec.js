import useGlobally from '../src/useGlobally'
import serialize from '../src/serialize'
import {MockPromise} from '../src/mockPromise'


test('Serialize matches snapshot', () => {
  expect(serialize({
    a: () => {}, 
    b: 'some text', 
    c: ['nested array'],
    d: {
      e: 'nested object'
    }
  })).toMatchSnapshot()

  expect(serialize(() => {})).toMatchSnapshot()
  expect(serialize('str')).toMatchSnapshot()
  expect(serialize(100)).toMatchSnapshot()
  expect(serialize(['txt', 10])).toMatchSnapshot()
  expect(serialize(new MockPromise('smth'))).toMatchSnapshot()
})


test('UseGlobally replaces global', () => {
  global.testVar = 10
  useGlobally('testVar', 11)

  expect(testVar).toBe(11)

  delete global.testVar
})