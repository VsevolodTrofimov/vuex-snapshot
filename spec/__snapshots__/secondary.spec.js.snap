// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serialize matches snapshot 1`] = `"{\\"b\\":\\"some text\\",\\"c\\":[\\"nested array\\"],\\"d\\":{\\"e\\":\\"nested object\\"}}"`;

exports[`Serialize matches snapshot 2`] = `undefined`;

exports[`Serialize matches snapshot 3`] = `"\\"str\\""`;

exports[`Serialize matches snapshot 4`] = `"100"`;

exports[`Serialize matches snapshot 5`] = `"[\\"txt\\",10]"`;

exports[`Serialize matches snapshot 6`] = `"{\\"name\\":\\"smth\\"}"`;

exports[`SnapState matches snapshot 1`] = `"{\\"b\\":\\"some text\\",\\"c\\":[\\"nested array\\"],\\"d\\":{\\"e\\":\\"nested object\\"}}"`;
